### PR TITLE
refactor(Alert): refactor closeIcon

### PR DESCRIPTION
**Pull request by kiner-tang**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [X] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- https://github.com/ant-design/ant-design/discussions/42828

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | - Alert support hidden close button when closeIcon setting to null or false <br />- deprecated closeText |
| 🇨🇳 Chinese | - 警告组件支持通过设置 closeIcon 为 null 或 false 隐藏关闭按钮 <br/>- 废弃 closeText 属性 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed
---

### <samp>🤖 Generated by Copilot at 2808e60</samp>

This pull request adds a new `closeIcon` prop to the Alert component, which allows users to customize or hide the close button of the alert. It also deprecates the `closeText` prop, which is replaced by the `closeIcon` prop. It updates the demo files, the test cases, and the API documentation to reflect these changes.

### <samp>🤖 Generated by Copilot at 2808e60</samp>

*  Add closeIcon prop to Alert component, allowing customization of the close button icon or hiding it by setting it to null or false ([link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-2aefe53da61f0b93281e271248872af3b04a0a74e56a10006ec813c710632dc6L23-R27), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-2aefe53da61f0b93281e271248872af3b04a0a74e56a10006ec813c710632dc6L44-R48), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-2aefe53da61f0b93281e271248872af3b04a0a74e56a10006ec813c710632dc6L87-R93), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-2aefe53da61f0b93281e271248872af3b04a0a74e56a10006ec813c710632dc6L139-R145), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-2aefe53da61f0b93281e271248872af3b04a0a74e56a10006ec813c710632dc6L203-R209), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-a3365a27d5160d6c7c1cf0364ef30876f602733ef443c734003ae750f1a8d104L13-R13), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-a3365a27d5160d6c7c1cf0364ef30876f602733ef443c734003ae750f1a8d104L20-R20), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-3dbaad52bcdfbd35c1528c90de61c60eaae0653da36bbd42dd88b6731e432248L1-R4), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-174ea1bf94f8ca33176522ecbdb934fbd907e6b53ae9307d94f4154fd774f561L43-R44), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-4493184779df5fd085712110fe4651b2cc37c39a689940232ca00af631d8544fL44-R45))
* Deprecate closeText prop in Alert component, logging a warning message when used and recommending to use closeIcon instead ([link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-2aefe53da61f0b93281e271248872af3b04a0a74e56a10006ec813c710632dc6R12), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-2aefe53da61f0b93281e271248872af3b04a0a74e56a10006ec813c710632dc6L114-R122), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-174ea1bf94f8ca33176522ecbdb934fbd907e6b53ae9307d94f4154fd774f561L43-R44), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-4493184779df5fd085712110fe4651b2cc37c39a689940232ca00af631d8544fL44-R45))
* Remove closeText prop from CloseIcon component, simplifying the props and the logic ([link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-2aefe53da61f0b93281e271248872af3b04a0a74e56a10006ec813c710632dc6L81), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-2aefe53da61f0b93281e271248872af3b04a0a74e56a10006ec813c710632dc6L87-R93), [link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-2aefe53da61f0b93281e271248872af3b04a0a74e56a10006ec813c710632dc6L203-R209))
* Add test cases for Alert component, checking the behavior of the close button when closeIcon is set to null or false, and the warning message when closeText is used ([link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-b3f47a1d74a43149ce487ead616f09aa09447b23a43513b97553f148f7b64474R144-R166))
* Reorder import statements in `closable.tsx` demo file, following the convention of importing React first ([link](https://github.com/ant-design/ant-design/pull/42962/files?diff=unified&w=0#diff-a3365a27d5160d6c7c1cf0364ef30876f602733ef443c734003ae750f1a8d104L1-R2))

### CODE REVIEW

File: components/alert/index.zh-CN.md [hunk 1]
```diff
@@ -41,9 +41,8 @@ group:
 | action | 自定义操作项 | ReactNode | - | 4.9.0 |
 | afterClose | 关闭动画结束后触发的回调函数 | () => void | - |  |
 | banner | 是否用作顶部公告 | boolean | false |  |
-| closable | 默认不显示关闭按钮 | boolean | - |  |
-| closeText | 自定义关闭按钮 | ReactNode | - |  |
-| closeIcon | 自定义关闭 Icon | ReactNode | `<CloseOutlined />` | 4.18.0 |
+| closeText | 自定义关闭按钮，>= 5.7.0 废弃，请使用 `closeIcon` 代替 | ReactNode | - |  |
```
Comment (zombieJ) [inline components/alert/index.zh-CN.md:44]: 废弃 API ，文档里移除

File: components/alert/index.zh-CN.md [hunk 2]
```diff
@@ -41,9 +41,8 @@ group:
...
+| closeIcon | 自定义关闭 Icon | boolean \| ReactNode | `<CloseOutlined />` | 5.7.0: 设置为 null 或 false 时隐藏关闭按钮 |
```
Comment (zombieJ) [inline components/alert/index.zh-CN.md:45]: `5.7.0: 设置为 null 或 false 时隐藏关闭按钮` 这个挪到说明里，放版本里太长了

File: components/alert/index.tsx [hunk 3]
```diff
@@ -136,8 +139,10 @@ const Alert: CompoundedComponent = ({
     return banner ? 'warning' : 'info';
   };
 
-  // closeable when closeText is assigned
-  const isClosable = closeText ? true : closable;
+  const closeIcon = oriCloseIcon !== null && oriCloseIcon !== false ? oriCloseIcon : null;
+
+  // closeable when closeText or closeIcon is assigned
+  const isClosable = closeIcon || closeText || closable;
```
Comment (zombieJ) [inline components/alert/index.tsx:145]: 这个顺序还是要保持兼容，closeText > closable > closeIcon。

File: components/alert/index.tsx [hunk 4]
```diff
@@ -78,16 +82,15 @@ const IconNode: React.FC<IconNodeProps> = (props) => {
...
   handleClose: AlertProps['onClose'];
 }
 
 const CloseIcon: React.FC<CloseIconProps> = (props) => {
-  const { isClosable, closeText, prefixCls, closeIcon, handleClose } = props;
+  const { isClosable, prefixCls, closeIcon = <CloseOutlined />, handleClose } = props;
   return isClosable ? (
     <button type="button" onClick={handleClose} className={`${prefixCls}-close-icon`} tabIndex={0}>
-      {closeText ? <span className={`${prefixCls}-close-text`}>{closeText}</span> : closeIcon}
+      {closeIcon === true ? <CloseOutlined /> : closeIcon}
```
Comment (zombieJ) [inline components/alert/index.tsx:93]: 这个写法有点坏味道...上面已经默认值是 `<CloseOutlined />` 了，这边又来一次 `<CloseOutlined />`。两个要合起来

**Review by RedJue — APPROVED** [review]

File: components/alert/index.tsx (hunk 4 above)
Comment (kiner-tang) [reply components/alert/index.tsx:93]: 当 closeIcon 是 true 的时候，不会走默认值，但有需要展示默认图标呀

File: components/alert/index.zh-CN.md [hunk 5]
```diff
@@ -41,8 +41,7 @@ group:
 | action | 自定义操作项 | ReactNode | - | 4.9.0 |
 | afterClose | 关闭动画结束后触发的回调函数 | () => void | - |  |
 | banner | 是否用作顶部公告 | boolean | false |  |
-| closeText | 自定义关闭按钮，>= 5.7.0 废弃，请使用 `closeIcon` 代替 | ReactNode | - |  |
-| closeIcon | 自定义关闭 Icon | boolean \| ReactNode | `<CloseOutlined />` | 5.7.0: 设置为 null 或 false 时隐藏关闭按钮 |
+| closeIcon | 自定义关闭 Icon，>=5.7.0: 设置为 null 或 false 时隐藏关闭按钮 | boolean \| ReactNode | `<CloseOutlined />` |  |
```
Comment (afc163) [inline components/alert/index.zh-CN.md:43]: 貌似没支持 closable？

Comment (kiner-tang) [reply components/alert/index.zh-CN.md:43]: 之前讨论之后确定吧 closable 从文档删除，但代码暂时不做处理，让用户平缓过度使用 closeIcon 隐藏

File: components/alert/index.en-US.md [hunk 6]
```diff
@@ -40,9 +39,7 @@ Alert component for feedback.
 | action | The action of Alert | ReactNode | - | 4.9.0 |
 | afterClose | Called when close animation is finished | () => void | - |  |
 | banner | Whether to show as banner | boolean | false |  |
-| closable | Whether Alert can be closed | boolean | - |  |
-| closeText | Close text to show | ReactNode | - |  |
-| closeIcon | Custom close icon | ReactNode | `<CloseOutlined />` | 4.18.0 |
+| closeIcon | Custom close icon, >=5.7.0: close button will be hidden when setting to null or false | boolean \| ReactNode | `<CloseOutlined />` |  |
```
Comment (zombieJ) [inline components/alert/index.en-US.md:42]: ``` ` null` ``` ``` `false` ``` 少了反引号

File: components/alert/__tests__/index.test.tsx [hunk 7]
```diff
@@ -141,4 +141,29 @@ describe('Alert', () => {
     const { container } = render(<Alert description="description" />);
     expect(!!container.querySelector('.ant-alert-message')).toBe(false);
   });
+
+  it('close button should be hidden when closeIcon setting to null or false', () => {
+    const { container, rerender } = render(<Alert closeIcon={null} />);
+    expect(container.querySelectorAll('.ant-alert-close-icon').length).toBe(0);
```
Comment (zombieJ) [inline components/alert/__tests__/index.test.tsx:147]:
```tsx
expect(container.querySelector('.ant-alert-close-icon')).toBeFalsy();
```

下同

File: components/alert/__tests__/index.test.tsx [hunk 8]
```diff
@@ -141,4 +141,29 @@ describe('Alert', () => {
...
+    rerender(<Alert closeIcon={false} />);
+    expect(container.querySelectorAll('.ant-alert-close-icon').length).toBe(0);
+    rerender(<Alert closeIcon />);
+    expect(container.querySelectorAll('.ant-alert-close-icon').length).toBe(1);
+    rerender(<Alert />);
+    expect(container.querySelectorAll('.ant-alert-close-icon').length).toBe(0);
+  });
+
+  it('should warning when using closeText', () => {
+    const warnSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
```
Comment (zombieJ) [inline components/alert/__tests__/index.test.tsx:157]: 上面加个 resetWarn() 重置一下

File: components/alert/index.tsx [hunk 9]
```diff
@@ -111,12 +116,12 @@ const Alert: CompoundedComponent = ({
...
   closable,
   closeText,
-  closeIcon = <CloseOutlined />,
+  closeIcon: oriCloseIcon,
   action,
   ...props
 }) => {
   const [closed, setClosed] = React.useState(false);
-
+  warning(!closeText, 'Alert', '`closeText` is deprecated. Please use `closeIcon` instead.');
```
Comment (zombieJ) [inline components/alert/index.tsx:124]: 外面包一个 process.env.NODE_ENV 检查，这个 warning 在 prod 下不能直接 uglify

File: components/alert/index.tsx [hunk 10]
```diff
@@ -136,8 +141,10 @@ const Alert: CompoundedComponent = ({
     return banner ? 'warning' : 'info';
   };
 
-  // closeable when closeText is assigned
-  const isClosable = closeText ? true : closable;
+  const closeIcon = oriCloseIcon !== null && oriCloseIcon !== false ? oriCloseIcon : null;
+
+  // closeable when closeText or closeIcon is assigned
+  const isClosable = closeText || closeIcon || closable;
```
Comment (zombieJ) [inline components/alert/index.tsx:147]: 这个要稍微兼容一下原本的逻辑：
```tsx
const isClosable = React.useMemo(() => {
  if (closeText) {
    return true;
  }

  if (!closable) {
    return false;
  }

  return ...
}, []);
```

File: components/alert/index.tsx [hunk 11]
```diff
@@ -200,9 +207,8 @@ const Alert: CompoundedComponent = ({
           {action ? <div className={`${prefixCls}-action`}>{action}</div> : null}
           <CloseIcon
             isClosable={!!isClosable}
-            closeText={closeText}
             prefixCls={prefixCls}
-            closeIcon={closeIcon}
+            closeIcon={closeIcon ?? closeText}
```
Comment (zombieJ) [inline components/alert/index.tsx:211]: 原来的逻辑了 closeText 优先级更高，这个要兼容：`closeText || closeIcon`

File: components/alert/index.tsx [hunk 12]
```diff
@@ -136,8 +143,19 @@ const Alert: CompoundedComponent = ({
...
-  const isClosable = closeText ? true : closable;
+  const closeIcon = oriCloseIcon !== null && oriCloseIcon !== false ? oriCloseIcon : null;
+
+  // closeable when closeText or closeIcon is assigned
+  const isClosable = React.useMemo(() => {
+    if (closeText) {
+      return true;
+    }
+    if (closable) {
+      return true;
```
Comment (zombieJ) [inline components/alert/index.tsx:154]: 这个我是个示例，closable === false 的时候也要归进去。`typeof closable === boolean` ...

File: components/alert/index.zh-CN.md (hunk 5 above)
Comment (afc163) [reply components/alert/index.zh-CN.md]: 没懂，统一方案不是 https://github.com/ant-design/ant-design/discussions/42828 么？

Comment (kiner-tang) [reply components/alert/index.zh-CN.md]: image
image

上周同步会上讨论的结果，是移除文档

File: components/alert/index.tsx [hunk 13]
```diff
@@ -136,8 +143,18 @@ const Alert: CompoundedComponent = ({
...
+  // closeable when closeText or closeIcon is assigned
+  const isClosable = React.useMemo(() => {
+    if (closeText) {
+      return true;
+    }
+    if (typeof closable === 'boolean') {
+      return closable;
+    }
+    // should be true when closeIcon is 0 or ''
+    return closeIcon !== false && closeIcon !== null && closeIcon !== undefined;
```
Comment (zombieJ) [inline components/alert/index.tsx:155]: `closeIcon` 默认不就是 `undefined` 么？

Comment (kiner-tang) [reply components/alert/index.tsx:155]: closeIcon 默认是 undefined，但在 Alert，默认不就是不展示关闭按钮的吗？在不传 closeText 和 closable 的情况下，如果 closeIcon 是 undefined ，不就应该不展示关闭按钮吗？

File: components/alert/index.zh-CN.md (hunk 5 above)
Comment (afc163) [reply components/alert/index.zh-CN.md]: 原因是啥，不打算支持 closable 了么？

Comment (kiner-tang) [reply components/alert/index.zh-CN.md]: closable={{ ... }} 形式书写比较麻烦，原本一层的变多层。
单个配置项的话 closeIcon={false} 直接就可以代替 closable={false} 的效果。
closeIcon 本身还带有配置 Icon 的能力，是一个超集。

得出结论，为了方便用户记忆和使用，统一  closeIcon=false 允许关闭，文档中不透出 `closable` 但是仍然支持。对于 Alert `closeText` 替换掉成 `closeIcon`。
`closable` 仍然支持 & 不 warning

**Review by zombieJ — APPROVED** [review]
